### PR TITLE
fix(cli): fix rudderstack anonymous ID generation in dev-client start command

### PR DESCRIPTION
**Pull request by wschurman**
# Why

We're seeing a spike in anonymous IDs for our metrics. This is probably due to `logEvent` being called (it's only called in one place) without having required auth first.

# How

The short-term fix proposed in this PR is to not log in cases where the user wasn't required to have auth'ed.

# Test Plan

1. `yarn build` in this package
2. `rm -rf ~/.expo` to clear any unique identifier previously generated.
3. In some expo app, run `nexpo start --dev-client`, see via temporary console.log that this would historically have used a random UUID for each time this is run. Now, see that it doesn't log.

Also run new test.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Review by EvanBacon — COMMENTED** [review]
Once we have the repro, should write a test for expected functionality since the logic here is a bit spaghetti

**Comment by kgc00** [comment]
The code looks good afaict, but we should confirm the behavior with a repro before merging.

File: packages/@expo/cli/src/utils/analytics/rudderstackClient.ts [hunk 1]
```diff
@@ -96,13 +98,20 @@ export async function logEventAsync(
     return;
   }
 
-  await getUserAsync().catch(() => null);
-  ensureIdentified();
+  // this has the side effect of calling `setUserData` which fetches the user and populates identifyData
+  try {
+    await getUserAsync();
```
Comment (tcdavis) [inline packages/@expo/cli/src/utils/analytics/rudderstackClient.ts:103]: I think in the case where the user is not authenticated, we'd skip setUserData and still log an event, but without the anonymous id set to anything?

Comment (kgc00) [reply packages/@expo/cli/src/utils/analytics/rudderstackClient.ts:103]: no id causes crashes on rudder server. we'd need to generate a random id or drop the event

**Comment by wschurman** [comment]
Updated test plan.

File: packages/@expo/cli/src/utils/analytics/__tests__/rudderstackClient-test.ts [hunk 2]
```diff
@@ -1,9 +1,37 @@
...
+  logEvent,
+  setUserDataAsync,
+} from '../rudderstackClient';
 
 jest.mock('ci-info', () => ({ isCI: true, isPR: true, name: 'GitHub Actions' }));
 
+jest.mock('@expo/rudder-sdk-node');
+
+describe(logEvent, () => {
+  // order of these tests matters since the module retains state
```
Comment (kgc00) [inline packages/@expo/cli/src/utils/analytics/__tests__/rudderstackClient-test.ts:15]: we should properly reset the state between tests. i think it's fine to leave this for a follow-up

**Review by kgc00 — APPROVED** [review]
seems like it will take care of the issue. we should keep an eye out over the next few days to see.

**Review by kgc00 — COMMENTED** [review]
On second look, am I understanding this correctly? It seems like calling `logEvent` will never identify the user. Shouldn't we replace these calls with the async version (which does identify the user) across the codebase in this PR.

File: packages/@expo/cli/src/utils/analytics/__tests__/rudderstackClient-test.ts [hunk 3]
```diff
@@ -1,9 +1,37 @@
...
+
+  it('does not log when the user has not been identified', () => {
```
Comment (tcdavis) [inline packages/@expo/cli/src/utils/analytics/__tests__/rudderstackClient-test.ts:17]: Maybe have a corresponding test for logEventAsync

**Review by tcdavis — APPROVED** [review]
Dropping the analytics works for me. Thanks so much for jumping on this!

It looks like the async calls might still be logging to me, but quite possible I'm not following and other folks are better equipped to sign off on that anyways.

**Review by kgc00 — APPROVED** [review]
seems like we just added the async event tracking for the start command. makes sense to me since that will generally be run before other commands.

File: packages/@expo/cli/src/utils/analytics/__tests__/rudderstackClient-test.ts (hunk 2 above)
Comment (byCedric) [reply packages/@expo/cli/src/utils/analytics/__tests__/rudderstackClient-test.ts:15]: Yeah, I think it should be isolated too. Not sure what order jest runs tests in 🤷

File: packages/@expo/cli/src/utils/analytics/rudderstackClient.ts [hunk 4]
```diff
@@ -96,27 +78,34 @@ export async function logEventAsync(
...
+  getRudderAnalyticsClient().track({
     event,
     properties: { ...properties, ...commonEventProperties },
     ...identity,
     context: getContext(),
   });
 }
 
-function ensureIdentified(): void {
+function identifyIfNotYetIdentified(): void {
```
Comment (EvanBacon) [inline packages/@expo/cli/src/utils/analytics/rudderstackClient.ts:103]: lol, any reason for the name change?

File: packages/@expo/cli/src/utils/analytics/rudderstackClient.ts [hunk 5]
```diff
@@ -96,27 +78,34 @@ export async function logEventAsync(
     return;
   }
 
-  await getUserAsync().catch(() => null);
-  ensureIdentified();
```
Comment (EvanBacon) [inline packages/@expo/cli/src/utils/analytics/rudderstackClient.ts:100]: This seems like it was rewritten but not changed

**Review by EvanBacon — APPROVED** [review]
Only thing that seems critical is the tests not reseting state when the explicit purpose of this PR is to ensure a user is identified.

**Pull request merged**